File: ios/LndMobile/LndUtils.swift
import Foundation

@objc(LndUtils)
class LndUtils: RCTEventEmitter {
  
  static let logEventName: String = "logEvent"
  
  var logEventsStarted: Bool = false
  
  static var lndDirectory: URL {
    let applicationDirectory = FileManager.default.urls(for: .applicationSupportDirectory, in: .userDomainMask).first!
    let lndDirectory = applicationDirectory.appendingPathComponent("lnd", isDirectory: true)
    
    if !FileManager.default.fileExists(atPath: lndDirectory.path) {
      try! FileManager.default.createDirectory(atPath: lndDirectory.path, withIntermediateDirectories: true)
    }
    
    return lndDirectory
  }
  
  static var confFile: URL {
    return lndDirectory.appendingPathComponent("lnd_v5.conf", isDirectory: false)
  }
  
  static private var logFile: URL {
    let network = Bundle.main.object(forInfoDictionaryKey: "NETWORK") as? String
    return lndDirectory.appendingPathComponent("logs", isDirectory: true)
      .appendingPathComponent("bitcoin", isDirectory: true)
      .appendingPathComponent(network ?? "mainnet", isDirectory: true)
      .appendingPathComponent("lnd.log", isDirectory: false)
  }
  
  @objc
  override static func moduleName() -> String! {
    "LndUtils"
  }
  
  override func supportedEvents() -> [String]! {
    return [LndUtils.logEventName]
  }
  
  @objc
  override static func requiresMainQueueSetup() -> Bool {
    return false
  }
  
  func writeConf(_ content: String) throws {
    try content.write(to: LndUtils.confFile, atomically: true, encoding: .utf8)
  }
  
  @objc(writeConf:resolver:rejecter:)
  func writeConf(_ content: String, resolver resolve: @escaping RCTPromiseResolveBlock, rejecter reject: @escaping RCTPromiseRejectBlock) {
    do {
      try self.writeConf(content)
      RCTLogInfo("conf file written")
      resolve("conf file written")
    } catch let err {
      RCTLogError("error writing conf: \(err.localizedDescription)")
      reject("error", err.localizedDescription, err)
    }
  }
  
  func writeDefaultConf() throws {
    let network = Bundle.main.object(forInfoDictionaryKey: "NETWORK") as? String
    var content = ""
    
    if network == "mainnet" {
      content =
        """
[Application Options]
debuglevel=warn
maxbackoff=2s
nobootstrap=1
nolisten=1
norest=1
sync-freelist=1
accept-keysend=1
feeurl=https://nodes.lightning.computer/fees/v1/btc-fee-estimates.json
gc-canceled-invoices-on-startup=1
gc-canceled-invoices-on-the-fly=1
ignore-historical-gossip-filters=1

[Bitcoin]
bitcoin.active=1
bitcoin.mainnet=1
bitcoin.node=neutrino

[Neutrino]
neutrino.addpeer=btcd-mainnet.lightning.computer
neutrino.addpeer=mainnet1-btcd.zaphq.io
neutrino.addpeer=mainnet2-btcd.zaphq.io
neutrino.addpeer=mainnet3-btcd.zaphq.io
neutrino.addpeer=mainnet4-btcd.zaphq.io
neutrino.addpeer=bb2.breez.technology
neutrino.validatechannels=false

[protocol]
protocol.option-scid-alias=true
protocol.zero-conf=true

[autopilot]
autopilot.active=0
autopilot.private=0
autopilot.minconfs=0
autopilot.conftarget=30
autopilot.allocation=1.0
autopilot.heuristic=externalscore:0.95
autopilot.heuristic=preferential:0.05
"""
    } else if (network == "testnet") {
      content =
        """
[Application Options]
debuglevel=info
maxbackoff=2s
nobootstrap=1
nolisten=1
norest=1
sync-freelist=1
accept-keysend=1
feeurl=https://nodes.lightning.computer/fees/v1/btc-fee-estimates.json
gc-canceled-invoices-on-startup=1
gc-canceled-invoices-on-the-fly=1
ignore-historical-gossip-filters=1

[Bitcoin]
bitcoin.active=1
bitcoin.testnet=1
bitcoin.node=neutrino

[Neutrino]
neutrino.connect=faucet.lightning.community
neutrino.validatechannels=false

[protocol]
protocol.option-scid-alias=true
protocol.zero-conf=true

[autopilot]
autopilot.active=0
autopilot.private=0
autopilot.minconfs=0
autopilot.conftarget=30
autopilot.allocation=1.0
autopilot.heuristic=externalscore:0.95
autopilot.heuristic=preferential:0.05
"""
    } else if (network == "regtest") {
      content =
        """
[Application Options]
debuglevel=info
maxbackoff=2s
nobootstrap=1
nolisten=1
norest=1
sync-freelist=1
accept-keysend=1
gc-canceled-invoices-on-startup=1
gc-canceled-invoices-on-the-fly=1
ignore-historical-gossip-filters=1

[Bitcoin]
bitcoin.active=1
bitcoin.regtest=1
bitcoin.node=bitcoind

[Bitcoind]
bitcoind.rpchost=127.0.0.1:18449
bitcoind.rpcuser=polaruser
bitcoind.rpcpass=polarpass
bitcoind.zmqpubrawblock=tcp://127.0.0.1:28340
bitcoind.zmqpubrawtx=tcp://127.0.0.1:29341

[protocol]
protocol.option-scid-alias=true
protocol.zero-conf=true

[autopilot]
autopilot.active=0
autopilot.private=0
autopilot.minconfs=0
autopilot.conftarget=30
autopilot.allocation=1.0
autopilot.heuristic=externalscore:0.95
autopilot.heuristic=preferential:0.05
"""
    }
    
    try self.writeConf(content)
  }
  
  @objc(writeDefaultConf:rejecter:)
  func writeDefaultConf(_ resolve: @escaping RCTPromiseResolveBlock, rejecter reject: @escaping RCTPromiseRejectBlock) {
    do {
      try self.writeDefaultConf()
      RCTLogInfo("conf file written")
      resolve("conf file written")
    } catch let err {
      RCTLogError("error writing conf: \(err.localizedDescription)")
      reject("error", err.localizedDescription, err)
    }
  }

  func deleteLog() {
    do {
      let fileManager = FileManager.default
      if fileManager.fileExists(atPath: LndUtils.logFile.path) {
        try fileManager.removeItem(atPath: LndUtils.logFile.path)
      }
    } catch let err {
      RCTLogError("Error deleting log: \(err.localizedDescription)")
    }
  }
  
  @objc(startLogEvents:rejecter:)
  func startLogEvents(_ resolve: @escaping RCTPromiseResolveBlock, rejecter reject: @escaping RCTPromiseRejectBlock) {
    if !self.logEventsStarted {
      let fileHandle = FileHandle(forReadingAtPath: LndUtils.logFile.path)
      
      DispatchQueue.main.async(
        execute: { [self] in
          NotificationCenter.default.addObserver(
            forName: FileHandle.readCompletionNotification,
            object: fileHandle,
            queue: OperationQueue.main,
            using:{ [self] notification in
              if self.bridge != nil {
                let data = notification.userInfo?[NSFileHandleNotificationDataItem] as? Data
                if data != nil && (data?.count ?? 0) > 0 {
                  let body = String(bytes: data!, encoding: .utf8)
                  self.sendEvent(withName: LndUtils.logEventName, body: body)
                }
                fileHandle?.readInBackgroundAndNotify()
              }
            })
          fileHandle?.seekToEndOfFile()
          fileHandle?.readInBackgroundAndNotify()
          self.logEventsStarted = true
          RCTLogInfo("log events started")
        }
      )
    }
    
    resolve("log events started")
  }
}

File: android/app/src/main/java/com/satimoto/LndUtils.java
package com.satimoto;

import android.os.FileObserver;

import android.util.Log;

import java.io.BufferedReader;
import java.io.File;
import java.io.FileInputStream;
import java.io.FileNotFoundException;
import java.io.IOException;
import java.io.InputStreamReader;
import java.io.PrintWriter;

import com.facebook.react.bridge.Promise;
import com.facebook.react.bridge.ReactApplicationContext;
import com.facebook.react.bridge.ReactMethod;
import com.facebook.react.bridge.ReactContextBaseJavaModule;
import com.facebook.react.modules.core.DeviceEventManagerModule;

public class LndUtils extends ReactContextBaseJavaModule {
    final String TAG = "LndUtils";
    private static final String logEventName = "logEvent";

    private FileObserver logObserver;
    final String confPath;
    final String lndPath;
    final String logPath;

    public LndUtils(ReactApplicationContext reactContext) {
        super(reactContext);

        lndPath = getReactApplicationContext().getFilesDir().toString();
        confPath = lndPath + "/lnd_v3.conf";
        logPath = lndPath + "/logs/bitcoin/" + BuildConfig.NETWORK + "/lnd.log";

        prepareFileDirectory(logPath);
    }

    @Override
    public String getName() {
        return TAG;
    }

    @ReactMethod
    public void addListener(String eventName) {}

    @ReactMethod
    public void removeListeners(Integer count) {}

    private void prepareFileDirectory(String filename) {
        new File(filename).getParentFile().mkdirs();
    }

    private PrintWriter getPrintWriter(String filename) throws Exception {
        prepareFileDirectory(filename);
        return new PrintWriter(filename);
    }

    void writeConf(String content) throws Exception {
        PrintWriter writer = getPrintWriter(confPath);
        writer.println(content);
        writer.close();
    }

    @ReactMethod
    void writeConf(String content, Promise promise) {
        try {
            this.writeConf(content);
            Log.d(TAG, "Saved LND conf to: " + confPath);
        } catch (Exception e) {
            Log.e(TAG, "Could not write to " + confPath, e);
            promise.reject("Could not write to : " + confPath, e);
            return;
        }

        promise.resolve("Saved LND conf to: " + confPath);
    }

    void writeDefaultConf() throws Exception {
        PrintWriter writer = getPrintWriter(confPath);

        if (BuildConfig.NETWORK.equals("mainnet")) {
            writer.println(
                    "[Application Options]\n" +
                            "debuglevel=warn\n" +
                            "maxbackoff=2s\n" +
                            "nolisten=1\n" +
                            "norest=1\n" +
                            "sync-freelist=1\n" +
                            "accept-keysend=1\n" +
                            "feeurl=https://nodes.lightning.computer/fees/v1/btc-fee-estimates.json\n" +
                            "tlsdisableautofill=1\n" +
                            "\n" +
                            "[Bitcoin]\n" +
                            "bitcoin.active=1\n" +
                            "bitcoin.mainnet=1\n" +
                            "bitcoin.node=neutrino\n" +
                            "\n" +
                            "[Neutrino]\n" +
                            "neutrino.addpeer=btcd-mainnet.lightning.computer\n" +
                            "neutrino.addpeer=mainnet1-btcd.zaphq.io\n" +
                            "neutrino.addpeer=mainnet2-btcd.zaphq.io\n" +
                            "neutrino.addpeer=mainnet3-btcd.zaphq.io\n" +
                            "neutrino.addpeer=mainnet4-btcd.zaphq.io\n" +
                            "neutrino.addpeer=bb2.breez.technology\n" +
                            "\n" +
                            "[protocol]\n" +
                            "protocol.option-scid-alias=true\n" +
                            "protocol.zero-conf=true\n" +
                            "\n" +
                            "[autopilot]\n" +
                            "autopilot.active=0\n" +
                            "autopilot.private=0\n" +
                            "autopilot.minconfs=0\n" +
                            "autopilot.conftarget=30\n" +
                            "autopilot.allocation=1.0\n" +
                            "autopilot.heuristic=externalscore:0.95\n" +
                            "autopilot.heuristic=preferential:0.05\n"
            );
        } else if (BuildConfig.NETWORK.equals("testnet")) {
            writer.println(
                    "[Application Options]\n" +
                            "debuglevel=info\n" +
                            "maxbackoff=2s\n" +
                            "nolisten=1\n" +
                            "norest=1\n" +
                            "sync-freelist=1\n" +
                            "accept-keysend=1\n" +
                            "feeurl=https://nodes.lightning.computer/fees/v1/btc-fee-estimates.json\n" +
                            "tlsdisableautofill=1\n" +
                            "\n" +
                            "[Bitcoin]\n" +
                            "bitcoin.active=1\n" +
                            "bitcoin.testnet=1\n" +
                            "bitcoin.node=neutrino\n" +
                            "\n" +
                            "[Neutrino]\n" +
                            "neutrino.connect=faucet.lightning.community\n" +
                            "\n" +
                            "[protocol]\n" +
                            "protocol.option-scid-alias=true\n" +
                            "protocol.zero-conf=true\n" +
                            "\n" +
                            "[autopilot]\n" +
                            "autopilot.active=0\n" +
                            "autopilot.private=0\n" +
                            "autopilot.minconfs=0\n" +
                            "autopilot.conftarget=30\n" +
                            "autopilot.allocation=1.0\n" +
                            "autopilot.heuristic=externalscore:0.95\n" +
                            "autopilot.heuristic=preferential:0.05\n"
            );
        } else if (BuildConfig.NETWORK.equals("regtest")) {
            writer.println(
                    "[Application Options]\n" +
                            "debuglevel=info\n" +
                            "maxbackoff=2s\n" +
                            "nolisten=1\n" +
                            "norest=1\n" +
                            "sync-freelist=1\n" +
                            "accept-keysend=1\n" +
                            "tlsdisableautofill=1\n" +
                            "\n" +
                            "[Bitcoin]\n" +
                            "bitcoin.active=1\n" +
                            "bitcoin.regtest=1\n" +
                            "bitcoin.node=bitcoind\n" +
                            "\n" +
                            "[Bitcoind]\n" +
                            "bitcoind.rpchost=10.0.2.2:18448\n" +
                            "bitcoind.rpcuser=polaruser\n" +
                            "bitcoind.rpcpass=polarpass\n" +
                            "bitcoind.zmqpubrawblock=tcp://10.0.2.2:28339\n" +
                            "bitcoind.zmqpubrawtx=tcp://10.0.2.2:29340\n" +
                            "\n" +
                            "[protocol]\n" +
                            "protocol.option-scid-alias=true\n" +
                            "protocol.zero-conf=true\n" +
                            "\n" +
                            "[autopilot]\n" +
                            "autopilot.active=0\n" +
                            "autopilot.private=0\n" +
                            "autopilot.minconfs=0\n" +
                            "autopilot.conftarget=30\n" +
                            "autopilot.allocation=1.0\n" +
                            "autopilot.heuristic=externalscore:0.95\n" +
                            "autopilot.heuristic=preferential:0.05\n"
            );
        }

        writer.close();
    }

    @ReactMethod
    void writeDefaultConf(Promise promise) {
        try {
            this.writeDefaultConf();
            Log.d(TAG, "Saved LND conf to: " + confPath);
        } catch (Exception e) {
            Log.e(TAG, "Could not write to " + confPath, e);
            promise.reject("Could not write to : " + confPath, e);
            return;
        }

        promise.resolve("Saved LND conf to: " + confPath);
    }

    @ReactMethod
    void startLogEvents(Promise promise) {
        if (logObserver == null) {
            FileInputStream fileInputStream = null;

            try {
                fileInputStream = new FileInputStream(logPath);
            } catch (FileNotFoundException fnfe) {
                Log.e(TAG, "Error initializing log events: " + logPath, fnfe);
                promise.reject("Error initializing log events: " + logPath, fnfe);
                return;
            }

            InputStreamReader inputStreamReader = new InputStreamReader(fileInputStream);
            BufferedReader bufferedReader = new BufferedReader(inputStreamReader);

            logObserver = new FileObserver(logPath) {
                @Override
                public void onEvent(int event, String file) {
                    if (event != FileObserver.MODIFY) {
                        return;
                    }

                    try {
                        emitLogEvent(bufferedReader);
                    } catch (IOException ioe) {
                        Log.e(TAG, "Error emitting log event", ioe);
                    }
                }
            };

            Log.i(TAG, "Log events started");
            logObserver.startWatching();
        }

        promise.resolve("Log events started");
    }

    private void emitLogEvent(BufferedReader bufferedReader) throws IOException {
        String lines = "";
        while ((lines = bufferedReader.readLine()) != null) {
            getReactApplicationContext()
                    .getJSModule(DeviceEventManagerModule.RCTDeviceEventEmitter.class)
                    .emit(logEventName, lines);
        }
    }
}